Configure Select Interaction with featureOverlay layer

rajadain opened the pull request:
## Overview

Adds the ability to configure a `Select` interaction with a `featureOverlay` layer. This layer is expected to already exist in the map. This is useful for when the user wants the selected items to be added to a pre-existing layer, rather than an unmanaged internal layer. This change is completely backwards compatible, and does not affect existing functionality at all.

Also adds documentation and an example for the new functionality.

Fixes #8064 

### Demo

![2018-08-12 07 21 53](https://user-images.githubusercontent.com/1430060/44001459-8f2d9b38-9e00-11e8-910b-ee54c4eeb3aa.gif)

### Notes

This is my very first PR to openlayers. I'm excited, but also humble. Please let me know if there are any stylistic / philosophical issues with the code, and if I should change things to move in a different direction.

I'm not sure why, but the "Feature Overlay" demo feels a hair slower than the default "Click" demo.

MatthijsBon commented on src/ol/interaction/Select.js:
```diff
@@ -291,6 +291,15 @@ class Select extends Interaction {
+   * @return {module:ol/layer/Vector} Layer.
+   * @api
+   */
+  getFeatureOverlay() {
```
Would it really be necessary to expose the featureOverlay? If it is managed by the user, there is already a reference somewhere. Since it was not exposed before, I  think it does not have to be exposed now.

rajadain replied:
Ah good catch, this was from an earlier draft. I'll remove it.

MatthijsBon commented on src/ol/interaction/Select.js:
```diff
@@ -149,7 +153,8 @@ class SelectEvent extends Event {
  *
- * Selected features are added to an internal unmanaged layer.
+ * Selected features are added to the `featureOverlay` layer if specified,
+ * otherwise in an internal unmanaged layer.
```
[..] "otherwise to an" [..]

rajadain commented:
Thank you so much for taking a look! I've dropped feature overlay getter commit, and fixed the typo:

```diff
diff --git a/src/ol/interaction/Select.js b/src/ol/interaction/Select.js
index 29aeb1a49..72b7217a4 100644
--- a/src/ol/interaction/Select.js
+++ b/src/ol/interaction/Select.js
@@ -154,7 +154,7 @@ class SelectEvent extends Event {
  * further feature filter using the `filter` option.
  *
  * Selected features are added to the `featureOverlay` layer if specified,
- * otherwise in an internal unmanaged layer.
+ * otherwise to an internal unmanaged layer.
  *
  * @fires SelectEvent
  * @api
@@ -302,15 +302,6 @@ class Select extends Interaction {
     return this.hitTolerance_;
   }
 
-  /**
-   * Returns the overlay layer that stores selected features.
-   * @return {module:ol/layer/Vector} Layer.
-   * @api
-   */
-  getFeatureOverlay() {
-    return this.featureOverlay_;
-  }
-
   /**
    * Returns the associated {@link module:ol/layer/Vector~Vector vectorlayer} of
    * the (last) selected feature. Note that this will not work with any
```

This is ready for another look.

MatthijsBon commented on src/ol/interaction/Select.js:
```diff
@@ -213,6 +213,12 @@ class Select extends Interaction {
+     * @private
+     * @type {boolean}
+     */
+    this.userSpecifiedFeatureOverlay_ = !!options.featureOverlay;
```
Personally I would be a bit more clear, e.q.   
`this.userSpecifiedFeatureOverlay_ = options.featureOverlay ? false : true;` 
but it has the same result. Even less obscure would be to do something like:  
```
this.userSpecifiedFeatureOverlay_ = typeof options.featureOverlay !== "undefined" 
                                    && options.featureOverlay !== null
```
Then the user would run into problems when he defines `options.featureOverlay` as something other than `null` or an actual `VectorLayer`, but that would be expected.

MatthijsBon replied:
Sorry for seeing this just now, otherwise the PR looks good!

rajadain replied:
Thanks for the pointer. With so many different ways of doing things, it's always great to be introduced to a project's style so the code looks homogenous. I amended the code as follows:

```diff
diff --git a/src/ol/interaction/Select.js b/src/ol/interaction/Select.js
index 72b7217a4..1c62ff3dc 100644
--- a/src/ol/interaction/Select.js
+++ b/src/ol/interaction/Select.js
@@ -217,13 +217,14 @@ class Select extends Interaction {
      * @private
      * @type {boolean}
      */
-    this.userSpecifiedFeatureOverlay_ = !!options.featureOverlay;
+    this.userSpecifiedFeatureOverlay_ = typeof options.featureOverlay !== 'undefined' &&
+      options.featureOverlay !== null;
 
     /**
      * @private
      * @type {module:ol/layer/Vector}
      */
-    this.featureOverlay_ = options.featureOverlay ? options.featureOverlay : new VectorLayer({
+    this.featureOverlay_ = this.userSpecifiedFeatureOverlay_ ? options.featureOverlay : new VectorLayer({
       source: new VectorSource({
         useSpatialIndex: false,
         features: options.features,
```

MatthijsBon replied:
No problem. I'm not really an OpenLayers pro, but since I added the issue, I thought I could review your PR. I think you copy pasted my comment code? Because now the indentation looks a bit off (sorry.. 😉 ) If that's fixed I'll approve it!

rajadain replied:
The indentation was actually a conscious choice, as I was trying to conform to this project's styling, for example see here:

https://github.com/openlayers/openlayers/blob/a94dff2c0635cce72f2ca6de32fe6805d26c0444/src/ol/interaction/Modify.js#L238-L243

I don't have a problem with indentation and can go either way, but would like to do that which is idiomatic to this project. Maybe a core developer can help us here.

ahocevar commented on src/ol/interaction/Select.js:
```diff
@@ -208,7 +213,18 @@ class Select extends Interaction {
+     * @private
+     * @type {boolean}
+     */
+    this.userSpecifiedFeatureOverlay_ = typeof options.featureOverlay !== 'undefined' &&
```
I'd prefer a simpler name, e.g. `this.userLayer_`. Also, please compute the boolean value with
```js
this.userLayer_ = options.featureOverlay !== undefined;
```
instead.

ahocevar commented on src/ol/interaction/Select.js:
```diff
@@ -321,11 +333,17 @@ class Select extends Interaction {
    */
   setMap(map) {
+    if (this.userSpecifiedFeatureOverlay_) {
+      return;
```
Don't bail out of the whole function. Only
```js
this.featureOverlay_.setMap(map);
```
should be executed conditionally.

ahocevar commented on src/ol/interaction/Select.js:
```diff
@@ -255,11 +265,13 @@ class Select extends Interaction {
-      this.addFeature_, this);
-    listen(features, CollectionEventType.REMOVE,
-      this.removeFeature_, this);
+    if (!this.userSpecifiedFeatureOverlay_) {
```
Instead of treating a user layer differently, you could unconditionally listen to the source's `ADDFEATURE` and `REMOVEFEATURE` events:
 ```js
const source = this.featureOverlay_.getSource();
listen(source, VectorEventType.ADDFEATURE, this.addFeature_, this);
listen(source, VectorEventType.REMOVEFEATURE, this.removeFeature_, this);
```

rajadain replied:
Hi! I added fixups for the requested changes. The first three work great. Unfortunately, this one doesn't so well. The changes for this are in a5f4690. Instead of behavior like shown in the GIF in the description, with this I see:

![2018-08-14 21 14 08](https://user-images.githubusercontent.com/1430060/44126560-0b355e20-a007-11e8-9916-b5a2270ac371.gif)

Something about skipping the features makes them not render, which is why I had excluded these when a user layer was specified. If you can suggest any other workarounds I'll be happy to try them out too.

rajadain commented:
Thanks for the detailed feedback! I'll have a chance to address them this evening.

marcjansen commented:
The code looked decent, @rajadain, looking forward to future contributions.
